server/src/main/java/io/crate/execution/dml/upsert/FromRawInsertSource.java

```diff
@@ -21,27 +21,148 @@
-import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.xcontent.DeprecationHandler;
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
 import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.index.mapper.StrictDynamicMappingException;
 
 import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
+import com.fasterxml.jackson.core.JsonParseException;
+
 public class FromRawInsertSource implements InsertSourceGen {
 
-    public BytesReference generateSourceAndCheckConstraintsAsBytesReference(Object[] values) throws IOException {
-        return new BytesArray(((String) values[0]));
+    private final DocTableInfo table;
+    private final TransactionContext txnCtx;
+    private final NodeContext nodeCtx;
+    private final InputFactory inputFactory;
+    private final CheckConstraints<Map<String, Object>, CollectExpression<Map<String, Object>, ?>> checks;
+
+    // holds table columns utilized as well as dynamically added columns of its lifetime
+    private final Map<String, Reference> refLookUpCache = new HashMap<>();
+
+    public FromRawInsertSource(DocTableInfo table,
+                               TransactionContext txnCtx,
+                               NodeContext nodeCtx,
+                               String indexName) {
+        this.table = table;
+        this.inputFactory = new InputFactory(nodeCtx);
+        this.txnCtx = txnCtx;
+        this.nodeCtx = nodeCtx;
+        this.checks = new CheckConstraints<>(
+            txnCtx,
+            inputFactory,
+            new FromSourceRefResolver(table.partitionedByColumns(), indexName),
+            table
+        );
     }
 
     @Override
-    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) throws IOException {
-        return JsonXContent.JSON_XCONTENT.createParser(
-            NamedXContentRegistry.EMPTY,
-            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
-            new BytesArray(((String) values[0])).array()
-        ).map();
+    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) {
+        String src = (String) values[0];
+        Map<String, Object> parsed = Map.of();
+        try {
+            var parser = JsonXContent.JSON_XCONTENT.createParser(
+                NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                new BytesArray(src).array());
+            parsed = parser.map();
+            if (parser.getTokenLocation().columnNumber < src.length()) {
+                throw new IllegalArgumentException("failed to parse `" + src.substring(parser.getTokenLocation().columnNumber) + "`");
+            }
+        } catch (IOException e) {
+            if (e instanceof JsonParseException) {
+                throw new IllegalArgumentException("failed to parse");
+            }
+        }
```

Comment:
This would ignore other types of IOException - shouldn't they bubble up?
And why does it re-raise the JsonParseException? I suspect the original exception would have a more informative error message.

Reply:
Thank you, I should have re-thrown for the rest of IOExceptions. For JsonParseException, I was trying to make the messages more consistent. Also, the original message had a bit too much,

```
cr> copy e1 from '/tmp/t.json' return summary;
+---------------------------------------------------------+--------------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| node                                                    | uri                | success_count | error_count | errors                                                                                                                                                                                                                                                                        |
+---------------------------------------------------------+--------------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"id": "AA-O2L6hTg272-WfB1cjkQ", "name": "Tiejer Flue"} | file:///tmp/t.json |             0 |           1 | {"com.fasterxml.jackson.core.JsonParseException: Unexpected character ('}' (code 125)): expected a value\n at [Source: (byte[])\"{\"a\":}\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\"; line: 1, column: 7]": {"count": 1, "line_numbers": [1]}} |
+---------------------------------------------------------+--------------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Reply:
Please check 5a7acee for updates